### ISSUE TITLE
Master Issue for Language Server Crash

### ISSUE THREAD

**Issue by octref** [octref/vetur#136]
First, sorry for the LS crashes.

Here is 0.5.6 in case people need to do work: [vetur-0.5.6.vsix.zip](https://github.com/octref/vetur/files/947389/vetur-0.5.6.vsix.zip).  
Instruction: https://code.visualstudio.com/docs/editor/extension-gallery#_install-from-a-vsix

I'll be traveling during weekend, so I don't think all crashes can be resolved until Monday.

With that being said, here is a guideline for reporting server crashes to help me better collect info and solve the issue:

Please provide the following info for crashes:

1. vetur version
2. error log
3. **a minimal repro case**, which
    - should be either a single file or a small project generated by https://github.com/vuejs/vue-cli
    - tweak it until it crashes vetur's LS
    - post the code on GH issue (for single file) or upload your project to GH and post link (for project)

3 is really important, because without a repro I can't do anything.

Also @sandersn, can you take a look at #127 #128 #131 #135? They seem to come from the module resolution code in your PR.

**Comment by sandersn** [comment]
#128 comes from the hidden, inserted `import Vue from 'vue'` conflicting with an `import Vue` written by the user. (Based on my knowledge of the code &mdash; I haven't reproed it yet.)

**Comment by sandersn** [comment]
#127 and #135 seem to be caused by incorrect setup of vetur's dependency on vue-template-compiler. It is likely that I didn't set up this dependency correctly, which is why @MWallin was able to work around the problem by installing vue-template-compiler as part of their own package.

**Comment by octref** [comment]
Yea I also see #128 coming from there. As for #135, @MWallin seems to be using `vue-template-compiler` as a devDeps and that somehow have an effect on vetur's LS.

**Comment by sandersn** [comment]
Looks like `package.json` at root needs to add a dependency on `vue-template-compiler` according to https://code.visualstudio.com/docs/extensionAPI/extension-manifest

**Comment by sandersn** [comment]
It's probably as simple as adding 

```json
  "dependencies": {
    "vue-template-compiler": "^2.2.1"
  },
```

in the middle of the root `package.json`. I'm not sure how to test it though.

**Comment by octref** [comment]
Also, think we should limit the files to do module resolution for limited to only `.vue` file.

> Looks like package.json at root needs to add a dependency on vue-template-compiler according to https://code.visualstudio.com/docs/extensionAPI/extension-manifest

Why so? I'm compiling `/server` into `/client`, and then publishing the whole thing. So `vue-template-compiler@2.2.1` should already be included in the LS.

**Comment by sandersn** [comment]
I was guessing based on the failing behaviour; it seems like user dependencies on `vue-template-compiler` (or lack of dependency) affects whether the extension crashes or not. vetur's installation of `vue-template-compiler` should be independent of whether the user project has a dependency on it. And adding it to the extension manifest (what VS Code calls the root package.json) seems like exactly the sort of extra-npm thing that would make the extension's dependencies separate.

That's my reasoning. It's conjecture but it seems worth testing at least.

I'll contact one of the VS Code guys and see if he can help. (I'll ping him too when I can find his github alias ....)

**Comment by sandersn** [comment]
> Also, think we should limit the files to do module resolution for limited to only .vue file.

Do you mean that only `import X from 'x.vue'` should work and not `import Y from 'someJavascript'`? What about things like `import _ from 'jquery'` and other common libraries?

**Comment by octref** [comment]
> That's my reasoning. It's conjecture but it seems worth testing at least.

I'll do that, package it as vsix and ask people who have problems to try it again.

> Do you mean that only import X from 'x.vue' should work and not import Y from 'someJavascript'? What about things like import $ from 'jquery' and other common libraries?

I meant this function https://github.com/octref/vetur/blob/master/server/src/modes/javascriptMode.ts#L89-L109

is even trying to resolve js files like `webpack.config.js`. As VSCode's js/ts LS already handles that, I think it makes sense if we just resolve modules from `.vue` files, instead of `.vue` and `.js` files. Things like `import $ from 'jquery'` should still work in script section of any vue file.

**Comment by sandersn** [comment]
The problem with that code path is that it has no real way to say "I don't handle these kinds of files", so it has [to do the fallback itself](https://github.com/octref/vetur/blob/master/server/src/modes/javascriptMode.ts#L93-L94):

```ts
        if (path.isAbsolute(name) || !isVue(name)) {
          return ts.resolveModuleName(name, containingFile, compilerOptions, ts.sys).resolvedModule;
        }
        else {
```

But `ts.resolveModuleName` should behave exactly the same way as the default implementation of `resolveModuleNames`, so there shouldn't be any change in the way that `import X from 'webpack.config.js'` is resolved. (I'll go look at the code and see if I can confirm that.)

**Comment by jonthegiant** [comment]
Any chance that this stacktrace is related to the issue?

```
[15792:0421/171353:INFO:CONSOLE(29)] "Connection is closed.: Error: Connection is closed.
    at ConnectionError.Error (native)
    at new ConnectionError (C:\Users\XXXXXX\.vscode\extensions\octref.vetur-0.6.1\client\node_modules\vscode-jsonrpc\lib\main.js:116:28)
    at throwIfClosedOrDisposed (C:\Users\XXXXXX\.vscode\extensions\octref.vetur-0.6.1\client\node_modules\vscode-jsonrpc\lib\main.js:481:19)
    at Object.sendNotification (C:\Users\XXXXXX\.vscode\extensions\octref.vetur-0.6.1\client\node_modules\vscode-jsonrpc\lib\main.js:522:13)
    at C:\Users\XXXXXX\.vscode\extensions\octref.vetur-0.6.1\client\node_modules\vscode-jsonrpc\lib\main.js:627:32
    at e.invoke (c:\Program Files (x86)\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:4:16423)
    at e.fire (c:\Program Files (x86)\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:4:20054)
    at e.cancel (c:\Program Files (x86)\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:4:22821)
    at e.cancel (c:\Program Files (x86)\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:4:23431)
    at ._oncancel (c:\Program Files (x86)\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:4:58954)", source: file:///C:/Program Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/electron-browser/workbench.main.js
(29)
```

**Comment by jonthegiant** [comment]
@octref 

Can do, just wasn't sure if this required it's own issue.

**Comment by octref** [comment]
@jonthegiant Haven't seen similar error before so it's worth one.

**Comment by octref** [comment]
@sandersn 

> import X from 'webpack.config.js' is resolved.

I meant the `containingFile` here only be `.vue` files.

```ts
resolveModuleNames(moduleNames: string[], containingFile: string): ts.ResolvedModule[] {
```

Requiring `.js` files in `.vue` files should certainly be supported.
But I don't see the use of, say, resolve the `require('path')` in a user's `webpack.config.js`.
I'm mainly worried since there are reports about CPU and ram usage https://github.com/octref/vetur/issues/127#issuecomment-296167511 and https://github.com/octref/vetur/issues/131 and my guess is they come from trying to do too much module resolution in large projects.

**Comment by sandersn** [comment]
Oh, I see. A user might have `giantFile.js` open in one buffer, which is using VS Code's instance of the TS language service. Then when they do `import X from 'giantFile'` in a `.vue` file, Vetur's instance of the TS language service will go off and *also* analyse `giantFile.js`.

**Comment by sandersn** [comment]
I think there may be some tsconfig settings to limit how deeply the compiler follows module resolution inside `node_modules`.

**Comment by MWallin** [comment]
Don't know if its relevant or not, but I didn't really install `vue-template-compiler` myself. It was a part of the webpack-simple CLI template. Might that perhaps be part of why its hard to isolate whats wrong, lots of different ways to setup your projects...

**Comment by exonuclease** [comment]
When i open the project folder contained node_modeules the error message is JavaScript heap out of memory and if i open the components folder the error message is start < 0.it is strange

**Comment by octref** [comment]
`start < 0` should be fixed with 0.6.2. Heap problem seems unresolved yet -- see #131.
